### ISSUE TITLE
Error thrown by predictInterval()

### ISSUE THREAD

**Issue by ulfelder** [jknowles/merTools#31]
When I run predictInterval(), I am getting the following error:

`Error in model.matrix(matrixForm, data = tmp)[1:nrow(newdata), , drop = FALSE] :`
`  subscript out of bounds`

I updated the package after the revisions you made last night, and I do not get this error when I run it on the iris/fakeiris example I used yesterday in the conversation that led to the addition of returnSims.

To no avail, I have tried:

 - Subsetting the training and test data to only the model variables prior to modeling and calling `predictInterval()`, respectively.

 - Checking to make sure the variables included in my call to `lmer()` were formatted the same way in the training and test data; they are.

 - Reducing n.sims from 1000 to 100.

There are no missing values in any columns in either training or test data sets, and all of the string-variable values for which my multilevel model includes random intercepts that appear in the test set also appear in the training set (i.e., there are no new levels).

Any idea why this might be happening, and how to fix it?

**Comment by carlbfrederick** [comment]
Have you tried passing it the data from the merMod itself (merMod@frame)?

I will take a look more in depth when I get in front of my computer.

**Comment by jknowles** [comment]
Without a reproducible example it might be tricky to diagnose -- but here are some questions I have that might help hone in:

1. Is the model you are using an `lmer` or a `glmer`? If it is generalized, what is the link function?
2. How many rows are you trying to predict? `predictInterval` probably shouldn't be used for > 1000 rows because I don't think it's optimized quite yet to do that efficiently. If you try the top 5 rows of your test data does it work?
3. Are you calling the function by loading the library first? There are some internal functions in `merTools` that are necessary, but not exported, so I haven't tested how it behaves if you try `merTools::predictInterval(merMod, n.sims = 100)`.
4. Do you have any of these kinds of funny terms in your model call: `sqrt(x)`; `I(x^3)`; `cos(x)`, etc.?

**Comment by jknowles** [comment]
My guess is the bug is here: 

```
tmp <- plyr::rbind.fill(newdata, trimModelFrame(merMod@frame))
```

If you run: 

```
tempdata <- merTools:::trimModelFrame(myModel@frame)
```
Do you get a valid data.frame back with more than 0 rows and more than 0 columns?

**Comment by ulfelder** [comment]
Thank you for taking a look at this.
I do not get the error if I use `predictInterval(..., newdata = [old data], ...)`.
1. I am using a simple `lmer` with one normally distributed, continuous fixed effect and random intercepts for a string variable with 32 unique values: `netscore ~ wiki_diff + (1 | home_team)`
2. I am trying to generate 1,000 simulations for 526 new cases (rows).
3. I am loading the library at the start of my script using `library(merTools)`.
4. No, as noted in 1, the model is quite simple.
Do you think the problem might come from the fact that the categorical variable for which I'm generating random intercepts is formatted as string/character in both the training and test sets, and the order in which the values appear differs across the two data sets? The `unique()` values match perfectly across the two sets, but I haven't set them as factors and specified the levels and order.

**Comment by ulfelder** [comment]
My model object is `mod0`. If I run `tempdata <- merTools:::trimModelFrame(mod0@frame)`, I get a data frame with 32 observations of 3 variables. My `home_team` variable has 32 unique vaues (NFL team names).

**Comment by jknowles** [comment]
It could be that the categorical variable is a string, but we should be pretty robust to that -- that's what the `trimModelFrame` function is trying to do -- create a model.matrix with all factor levels in the training and test data. 

This is a bit strange -- let me debug a bit more.

**Comment by ulfelder** [comment]
If it would help to have the script and .csvs of the data I'm using, I would be happy to send them. The csvs are both under 100KB.

**Comment by jknowles** [comment]
I tried to reproduce the problem with the dummy data below -- can you check if this script works? If it does not work for you, I'll scratch my head. If it does, then it might help to see your script and data to debug the exact problem:

```
possNames <- expand.grid(letters,LETTERS)
possNames <- paste(possNames[, 1], possNames[, 2])

newFac <- sample(possNames, 32) 

modData <- data.frame(
  y = rnorm(500),
  x = rnorm(500),
  team_name = sample(newFac, 500, replace = TRUE)
)

library(merTools)
m0 <- lmer(y ~ x + (1|team_name), data = modData[1:250,])
testPreds <- predictInterval(m0, newdata = modData[251:500,])

```

**Comment by ulfelder** [comment]
That script works for me.

**Comment by jknowles** [comment]
Alright that's good. Can you send me the script/data to the e-mail linked to my GitHub account? Should be a relatively quick debug for me from there I hope.

**Comment by jknowles** [comment]
OK I've diagnosed it. I will add a fix to the function, but for now you need to append a column to the new data with your dependent variable filled with meaningless data. 

```
modData$y <- 1
testPreds <- predictInterval(m0, newdata = modData[251:500,])
```

The way I normalized the coefficient matrices was assuming that the data you wanted to predict would already have a dependent variable column and it would be well-defined. Without that, it is dropping all cases without a dependent variable -- *which kind of defeats the point of naming it predict*. This is an easy fix, and this was a great use case for us to test!
